### ISSUE TITLE
Release plans ?

### ISSUE THREAD

**Issue by vladak** [universal-ctags/ctags#1038]
Are there any plans to do a (stable) release ?

**Comment by masatake** [comment]
Could you see "Milestones" menu?
It is just my plan.

**Comment by vladak** [comment]
Yes, I see couple of versions there. Any concrete timelines ?

**Comment by masatake** [comment]
I would like to release a stable version than can be included in GNU/Linux distro with 2016. However, I cannot promise it as usual.

**Comment by hauleth** [comment]
Could we have some pre-release versions **tagged** on GitHub? It would make life so much easier for almost everyone who want to introduce Universal Ctags as a package.

**Comment by hauleth** [comment]
Ping?

**Comment by masatake** [comment]
I think it is far from releasing.

**Comment by hauleth** [comment]
@masatake I am not talking about full 1.0.0 release, but tagged versions like 0.0.1 (SemVer mention that anything that is 0.x.y is meant to be highly unstable) which would help with creating development packages for various package managers. This would also help with generating version as `git describe` would return current version in quite nice format.

**Comment by masatake** [comment]
Do you really need a tagged version?

https://github.com/universal-ctags/ctags/issues/1272

**Comment by AMDmi3** [comment]
Yes, we do really need a tagged version, otherwise there's a mess with versions and no way to know if the package is up to date.

https://repology.org/metapackage/ctags/information

**Comment by emptyflask** [comment]
[Homebrew](https://github.com/Homebrew/brew) requires tagged releases, and once universal-ctags is included in the core list, it can be used with other packages like [GNU Global](https://github.com/Homebrew/homebrew-core/pull/8309)...

**Comment by masatake** [comment]
We are working hard for releasing.

The reason I don't do tagging is I want "standard users" not to use broken version of universal-ctags.
Here "standard users" mean people who just use "binary" executable getting via a package system or something. I myself such a people about other software than ctags, I don't want to make them unhappy.

**Comment by cweagans** [comment]
@masatake I think people generally know that if it's a 0.x release, there will be some rough edges. I would really like to encourage you to regularly tag 0.x releases until universal ctags is to a point where you think "normal" users will be happy. This will really help with getting packaging taken care of before the 1.0.0 release.

**Comment by hauleth** [comment]
Excerpt from Server specification:

> Major version zero (0.y.z) is for initial development. Anything may change at any time. The public API should not be considered stable.

**Comment by anthraxx** [comment]
@masatake how about an xmas present in form of a release? :yum:

**Comment by masatake** [comment]
Sorry, but we have to do two things: stabilizing the file format called tags format v3 and writing its document. I thought I had much time for working on these topics in this fall. However, I could not find much time. The next Easter is the target.

**Comment by hauleth** [comment]
We aren’t talking about version 1.0 but about 0.0.1-alpha.1.pre right now. 

--
Łukasz Jan Niemier

**Comment by cweagans** [comment]
@masatake Whether you tag a release or not, the fact is that people are using universal ctags right now because it's already way better than exuberant ctags. I understand if you don't have the time or interest in maintaining these releases, but is that something you can delegate? I'm willing to step in and do some limited support for 0.x releases to help offload some of the maintenance burden. 

Please consider taking a page out of Neovim's book: they are releasing 0.x releases as they continue to improve the vim codebase. There is no expectation of backward compatibility in these releases, but they are extraordinarily useful for packaging, testing, daily use, etc.

**Comment by masatake** [comment]
Thank you for your proposal.

What I'm afraid of is getting many bug reports. We changed the default file format. It may have a big impact.
I think we cannot respond to the bug reports well because even I don't know well about the new file format. 

* introducing version 3 format,
  - making --format=3 as the alias for --output-format=u-ctags.
  - [rintiing TAG_FILE_FORMAT 3 when the options are give.
* update readtags to handle TAG_FILE_FORMAT 3, and
* documentation for the format

are needed.
Much of things are done in a pull request submitted by @b4n.

**Comment by cweagans** [comment]
> What I'm afraid of is getting many bug reports.

Just for the record, I think it's _totally okay_ to just close those with a canned response about 0.x backward compatibility not being supported. The answer is simply "Regenerate your tags file using the new format."

I will happily and ruthlessly close those bug reports if it means there's a tagged release to work off of.

**Comment by masatake** [comment]
> Just for the record, I think it's totally okay  to just close those with a canned response about 0.x backward  compatibility not being supported. The answer is simply "Regenerate your  tags file using the new format." 

You don't understand the point. ctags is a just backend tool.  The new format we are testing introduces backslash for escaping. As the result, client tools(==frontend tools) will not work expectedly. Regenerating
tags in the feature doesn't solve the issue. Rewriting or updating the tools are needed.

I can image users get angry with me for such format change. The people  may say "compatibility" is the most important. If I am an user, I also think so. So I'm trying "breaking compatiblity once, but only once when releasing 6.0." (if possible).

**Comment by masatake** [comment]
Anyway I will work on the format 3 issue. I must stop hacking fortran parser. It is written well.

**Comment by cweagans** [comment]
If you're strongly opposed to this, I'm probably not going to change your mind here. However, even if tooling needs to be updated to work with the new format, a pre-release would help the tooling maintainers accomplish that task.

The worst possible scenario here is that 6.0 is tagged and none of the tooling is updated. That would be really, really bad. The best case scenario is that lots of users try the pre-release and find legit bugs (not related to the tag format), get tooling updated, etc, and then the 6.0 release is an awesome success with associated tooling working perfectly on day 0.

I know what I would pick, but then again, I guess I'm not the maintainer here.

**Comment by codebrainz** [comment]
What about doing:

```
git tag -a v5.9-experimental-unstable-do-not-use \
  -m "An unsupported experimental unstable alpha version tag"
```

**Comment by masatake** [comment]
Could you review my pull request, #1695 ?

**Comment by masatake** [comment]
Could anyone review the change about man pages in #1737 ?
Updating the man pages is must for releasing.

**Comment by mwean** [comment]
Now that you merged #1737 are you open to tagging a release? 🙏

**Comment by masatake** [comment]
#1737 is not merged yet. Much more manual page updates are needed for releasing.

**Comment by blueyed** [comment]
Just for reference:

- project: https://github.com/universal-ctags/ctags/projects/8
- milestone(s): https://github.com/universal-ctags/ctags/milestones

**Comment by Eluminae** [comment]
Hello all ! Some time has passed. The lack of release seems to me a debatable choice. Anyway, this really prevent some distro packaging... 
https://github.com/void-linux/void-packages/issues/15689#issuecomment-545352098

Is it possible to have some sort of 6.0.0-rc1?

Really sorry if I seem like the annoying bee. I really don't know all the context and I probably say shit..

**Comment by docwhat** [comment]
I would argue that at this point, releasing *anything* would be more beneficial than waiting for a perfect release.

A release would:

- Raise awareness of the project and why it is awesome.
- Encourage participation.
- Publicly show that the project is here and improving.
- Make everyone feel better. ;-p

**Comment by k-takata** [comment]
I personally agree that releasing anything is better than a perfect release.
> "Done is better than perfect." -- Mark Zukerberg

Universal ctags project stated at 2015 and we have developed for four years. It means that some people might wait a release for four years ;-)
Even `6.0.0-alpha1` should be better than waiting forever.
We may notice what are needed for a stable release by releasing an alpha release.

I'm thinking what are needed for an alpha release.

* Project: [Releasing 6.0](https://github.com/universal-ctags/ctags/projects/8)
  I have already created a PR (#2259) for adding the items listed here.
  All public options should be written in the man page, but the description doesn't need to be perfect.
* Milestones: [M0 for 6.0](https://github.com/universal-ctags/ctags/milestone/8), [Mn for 6.0](https://github.com/universal-ctags/ctags/milestone/9)
  At least, incompatible behaviors should be written in the documents or the man pages. However, I don't think all the items listed here should be fixed before the alpha release. (Even some of them might be able to postpone to the next stable release (i.e. `6.1.0` or something).)
* Make `make distcheck` work.
  This command is used for creating a release package. I have already fixed by #2298.

We have already introduced some major incompatible changes: changed configuration files from `.ctags` to `.ctags.d/*.ctags`, changed the tags file format to support special characters (which caused problem on Windows, but already fixed with #2199), etc.
Does anyone want to introduce some more incompatible changes?
Major incompatible changes should be included in the alpha release, but other changes can be postponed.

Basically, I think we should focus on writing documents about incompatible behaviors.

**Comment by masatake** [comment]
I'll make a reply till next Sunday.

**Comment by masatake** [comment]
I must not ignore what a co-developer in the team says.

I promise that I focus on the tasks related to releasing.
However, I will not make a release until the condition that I will explain later is satisfied.

I'm thinking about just releasing 6.0.0 instead of releasing the alpha version.
Once we ship 6.0.0, I will release bugfix versions very frequently: 6.0.1, 6.0.2,...
Releasing-daily is one of my ideas as @k-takata does in ctags-win32.

As I will write, I have not thought about releasing for long days, I cannot enumerate tasks for releasing quickly. However, I can categorize tasks into four groups:

(1) documentations especially about incompatibility, 
(2) revising error messages and help messages, 
(3) making decisions about spec (e.g., what kind of field we use for representing return type of function.), and
(4) fixing a crash trigged by broken input.

I will focus on them instead of introducing new features.

However, I will not make a release until we can find one more long-term-contributor who is good at reviewing the documents we will write.

----
 
Do you know the goal (or objective) of this project?
When we took over this repository from @fishman, I had a chance to define it. You can find it in README.md.

> universal-ctags has the objective of continuing the development from what existed in the Sourceforge area. Github exuberant-ctags repository was started by Reza Jelveh and was later moved to the universal-ctags organization.

> The goal of the project is preparing and maintaining common/unified working space where people interested in making ctags better can work together.

I was shocked at the development of e-ctags became inactive when I started hacking ctags.
e-ctags projects produced impressive software. However, it is mostly dead. None can take over the project.

I thought we should not repeat it. So in this project, an UNOFFICIAL fork of e-ctags, I set the above goal and objective. I can reword them "making the development sustainable."
What I did first is writing test harness called "units," which you know very well. "units" helps us to keep the quality of parser without knowing the detail of the target language. It is for increasing sustainability. 

This project succeeds mostly; I convince someone will take over the development of u-ctags even if I die today. Of course, you are the primary candidate of the "someone" these days:-) 
Keeping development sustainability is more important than being popular or releasing.

 What I'm afraid of is my burning out. As this issue shows, so many people want a release.
(e-)ctags is a really big name, so once we make a release, many people may use the version of u-ctags.
Getting massive feedback, including bug reports, may kill us. Not-releasing makes u-ctags popular very slowly. So we can get such feedback incrementally. For now, I'm comfortable with the pace of receiving feedback. Most of all case, I can take action for a bug report in a week.

I know well-written documents help handle the (imaginary) massive feedback.
e.g., FAQ is a typical one we should have.
I can quote an item from FAQ in responding to message to reports.

We need a person assisting us in such an aspect. I thought we could found such a person easily because I could get not a few good developers like you. It seems that I was too optimistic.

See https://github.com/universal-ctags/ctags/issues/1038#issuecomment-411642880. The reporter uses the "pray" character.  As I always do, I ask the one to review documents (in https://github.com/universal-ctags/ctags/issues/1038#issuecomment-411647300). To know how one strongly needs a release, I have asked reports reviewing. 

I decided I ignore requests for releasing from "just-thumbing-up people" till I can find a person helping us writing documents actively.  If I find such a person, I will make a release, even none requests.
(I strongly believe "just-thumbing-up" doesn't contribute to extending sustainability.

Three days ago, I got a request for releasing from you. You contribute to extending sustainability. I cannot ignore your comment. So I will focus on the tasks for releasing.
However, let's wait for one more person to the team for releasing constantly.

I know you can help me writing documentation as you have done ever. However, I guess even you are not free from the burning-out.

----

If you (other than @k-takata) are interested in ctags-6.0.0, try to add a section "Multi-line pattern match" to man/ctags-optlib.7.rst.in and make a pull request.  You can start with http://docs.ctags.io/en/latest/optlib.html#multi-line-pattern-match.
You can also fix my broken English and make pull requests in the man pages of u-ctags.

**Comment by vladak** [comment]
@masatake I think I understand your concerns, helping to run comparatively small project with very limited number of core contributors. However, the thing is Universal things had become de facto **the** ctags implementation, no matter if you want it or not. So, producing releases may not change much. After all these days everyone seems to be following the continuous integration model so releasing often is not a such a big deal even for projects in alpha/beta stage. The OpenGrok project uses Universal ctags as one of its main inputs. Recently things broke because of changes in Universal ctags (not a bug per se, just different error handling, see the above reference) and I'd like to be able to tell the users to stick to certain version of Universal ctags, not a changeset ID. After all, if I remember correctly the semantic versioning model suggests to version a project at least 1.0.0 if it is used in production already and this is definitely the case of Universal ctags.

As for burning out, this is a real concern that is hard to deal with if you care about the project. Some detachment is necessary I think and then it will be a matter of self regulation - "want a feature ? Send a PR in !" (and there is the Github issues bot that can close inactive issues automatically after some time if you want to manage the workload defensively)

**Comment by masatake** [comment]
I want to write about OpenGrok first. I'm a user of OpenGrok. 
I would like to recommend you to make OpenGrok incorporate u-ctags. git-subtree may help you.
u-ctags incorporates packcc, a compiler-compiler with git-subtree. It works quite well.
As a result, a user of OpenGrok doesn't have install u-ctags separately.

Let me explain about the change behind the error. The change introduced to let users know that "the parser you want is already implemented in u-ctags."  I guess people define a parser because a parser for the language is not implemented in u-ctags at the time. If u-ctags has one as a built-in, people may want to use the built-in parser instead of the user-defined one. The fatal error may help the user known the built-in parser. If the user is not satisfied with the built-in parser, the user can (1) make the user-defined one override the built-in one with  `--_pretend-<NEWLANG>=<OLDLANG>` option, or make a pull request for improving the built-in parser.

Let me reply about your comment for releasing.
> However, the thing is Universal things had become de facto the ctags implementation, no matter if you want it or not. So, producing releases may not change much.

I agree with you, and this is the expected project-growing story I assumed. As I wrote about git-subtree, in many cases, the releasing is not needed, I think. (but I'm working hard now for releasing because I promised.)

Many people want us to make a release. I am not interested in the reason.
However, I recognized there were substantial needs for a release. I want to employ this chance maximally for making ctags project sustainable.

I AM NEGOTIATING WITH YOU.

If you seriously want a release, you have to join this project as a leader for improving the documents (the man pages and web pages).

U-ctags is a project developing a free (and/or opensource) software. So I cannot/should not forces users to contribute to the development, of course.

However, as I wrote repeatedly, I want to have one more eager member of the team to make the project sustainable. "making a release" is a hostage. I can use this card only once. 

**The criminal declaration:**

> Provide your time and skill!,  I will exchange them with universal-ctags-6.0.0.tar.gz.
> Add a section "Multi-line pattern match" to man/ctags-optlib.7.rst.in and make a pull request. This is the first step for the release you want. 
> You can start with http://docs.ctags.io/en/latest/optlib.html#multi-line-pattern-match.
> Don't contact the police.

I would like you to skip this comment if you are already busy with another free and/or opensource project. "I-can-work-on-one-more-project"  can be a cause of burn-out.

This is my second experiment in open-source development.
(See [here](http://people.redhat.com/yamato/talks/recruitingsource.pdf) if you are interested in the first experiment called recruiting source.) Let's watch this issue.

Again, I'm working for releasing.  I'm revising CLI together with the man pages.

**Comment by masatake** [comment]
Could you review man/ctags-client-tools.7.rst.in in #2459? (in English grammer aspect)

**Comment by masatake** [comment]
The first person comes here for the negotiation :-)

**Comment by masatake** [comment]
Could you review man/ctags.1.rst.in in #2573 (in English grammer aspect)?

**Comment by emptyflask** [comment]
#2573 looks good to me, but it looks like I'm too late, it's merged already.

**Comment by jwhitley** [comment]
Agreed w/ @emptyflask, #2573 English looks good.

**Comment by masatake** [comment]
@emptyflask, @jwhitley, thank you for taking cares.

**Comment by masatake** [comment]
#2573 is merged.

**Comment by vt-alt** [comment]
I have two important questions. Page https://ctags.io/ says that `Here’s a few specific things you can do to help us: • You can help us package ctags for different platforms.` But, there is no releases for years and years.

1. From what commit should I create package for our distro (ALT Linux)?

Also, @masatake says:

> We are working hard for releasing.
> 
> The reason I don't do tagging is I want "standard users" not to use broken version of universal-ctags.
> Here "standard users" mean people who just use "binary" executable getting via a package system or something. I myself such a people about other software than  ctags, I don't want to make them unhappy.

2. Is it still broken? I really don't want to create package with a broken ctags.

**Comment by masatake** [comment]
Basically, you don't have to make a commit for the git repository of ctags.
You can do as you want in your side. 

I have no plan to make a release. I made many explanation in this issue.
I think the best way to prepare a binary package shipping it daily (, weekly, or month, as you want) as far as there is a change in the git repo.
For win32, https://github.com/universal-ctags/ctags-win32 does so.
I recommend you to run make check in your platform before making a package.
If you need to be a member of universal-ctags organization for making the periodic releases for Alt, let me know.
I don't know what kind of technique is used in ctags-win32, but it will be worth to study it. 

If you want to keep t the quality of package for Alt, you can make a pull request for adding Alt as a test enviroment to the test run at Circleci or Travis-Ci. See https://github.com/universal-ctags/ctags/blob/master/circle.yml e.g.


> Is it still broken? I really don't want to create package with a broken ctags.

YES. there are so many issues have been reported. I know more issued.
However, it seems that many peeple use u-ctags already. It is broken but useful.

**Comment by masatake** [comment]
How do you do? I believe you are healthy.

Could you review man/ctags.1.rst.in in https://github.com/universal-ctags/ctags/pull/2579/commits/36aa4a333263b6654b94f8fa14ade3843444c1c5#diff-0f58ec26841e4489a3d0a370f50797c0 of  #2579 (especially in English grammer aspect)?

I explained --roles-<LANG>.<KIND> option. If you think I should add more examples, let me know.

**Comment by vt-alt** [comment]
@masatake
Very important question for me - can you please answer: _Will there be regression if we install u-ctags over exuberant ctags?_

> However, it seems that many peeple use u-ctags already. It is broken but useful.

They may be not aware of the scale of u-ctags brokeness.

**Comment by masatake** [comment]
> Will there be regression if we install u-ctags over exuberant ctags?

Yes. e.g.  u-ctags doesn't read ~/.ctags file though e-tags reads it.

**Comment by vt-alt** [comment]
Well, I think it's unacceptable to create package with knowingly heavily broken software. Lets wait until you say it's not broken or release.


ps. JFYI: See the versioning mess you caused by not assigning upstream version to u-ctags. Packages for distros **required** to have version. See how distros version differently universal-ctags package in various distros*:

**0.r6046.e65e91a2	20200111	20191203	201903312	20170704	20161028	2020.07.01 2020-07-01	2019-07-30	62f0144	5.8.20190415	0.20191025	0.0.20190412	0 0.0.0-0.0c78c0c	0_git20200205	0_git20191207	0git20191013	0git20190715 0git20181215	0git20180225	r20200106	r7253.7492b90e	g20200606	g20200502 e216bb44cb45c9ba4e741bf1d7a0e7241e4db3c9**

* Data from https://repology.org/project/universal-ctags/versions

This also means that end users will have hard time understanding which version they actually use.

That fact that you not release (not even plan to), **but** invite people to package u-ctags **already** creates releases for each of these distros. All these bug reports that you so fear to receive, as a consequence of release, is already go into bug trackers of the distros and up to this github issue pages.

**Comment by polyzen** [comment]
> If you need to be a member of universal-ctags organization for making the periodic releases for Alt, let me know.

This level of access is unnecessary for downstream packaging.

**Comment by masatake** [comment]
I should not invite people to package u-ctags.
I withdrew the message. https://github.com/universal-ctags/universal-ctags.github.io/pull/21

"periodical packaging" will be the best way as far as I maintain u-ctags.
But, I should not say much about making a package.

**Comment by masatake** [comment]
> This level of access is unnecessary for downstream packaging.

Thank you for the comment... So I wonder why people doesn't do like ctags-win32.

**Comment by masatake** [comment]
People who reviews the man pages, please, let me know how strongly you need a release.

**Comment by kalkin** [comment]
@masatake IMHO If I were you I wouldn't care about bug reports. If it works for you release it. If you get bug reports, fix them when you have time, if you don't have time, then people should deliver code or shut up.

**Comment by masatake** [comment]
How do you do? I believe you are healthy.

Could you review man/ctags.1.rst.in in https://github.com/universal-ctags/ctags/pull/2579/commits/36aa4a333263b6654b94f8fa14ade3843444c1c5#diff-0f58ec26841e4489a3d0a370f50797c0 of  #2579 (especially in English grammar aspect)?

I explained --roles-<LANG>.<KIND> option. If you think I should add more examples, let me know.

**Comment by jwhitley** [comment]
> Could you review man/ctags.1.rst.in in [36aa4a3#diff-0f58ec26841e4489a3d0a370f50797c0](https://github.com/universal-ctags/ctags/commit/36aa4a333263b6654b94f8fa14ade3843444c1c5#diff-0f58ec26841e4489a3d0a370f50797c0) of #2579 (especially in English grammar aspect)?

The English in this change looks good to me.

**Comment by masatake** [comment]
@jwhitley, thank you very much.

**Comment by ploxiln** [comment]
(Summary: just make a tag, call it an alpha.)

exuberant-ctags 5.8 is now over 10 years old, and does not build successfully on modern macOS without a minor patch, which I contributed back in https://github.com/Homebrew/homebrew-core/pull/42147 after briefly explaining that there is no way to get any ctags project to make any releases with fixes. (You can get fixes here, you just can't get any releases anywhere, for over a decade now.) It was while trying to install good old ctags a year ago that I found myself asking "is there some currently maintained ctags somewhere? there has to be..." and found this, universal-ctags. This is it. This is the only maintained ctags, and it's better than exuberant-ctags in every way, as far as I can tell. (I see there is #2356 but I didn't notice that on my own, and wouldn't have expected that to work.) I only use it with Vim, while working on C and Python, and for this purpose it works surprisingly well. Literally the only thing it needs for the common ctags user, is a version number. (Most programmers these days use IDEs which do not use ctags or anything like it, they use custom built-in parser/analyzer/completion, or a "language server protocol" daemon. I'm sure there are other ctags use-cases out there but they're not going to be nearly as common as good old vim and emacs.)

This is your repo, your project, and your work, and I respect that. You've explained your anti-release rationale, which I will briefly quote:

> Many people want us to make a release. I am not interested in the reason. However, I recognized there were substantial needs for a release. I want to employ this chance maximally for making ctags project sustainable.

> I AM NEGOTIATING WITH YOU. If you seriously want a release, you have to join this project as a leader for improving the documents (the man pages and web pages).

> U-ctags is a project developing a free (and/or opensource) software. So I cannot/should not forces users to contribute to the development, of course. However, as I wrote repeatedly, I want to have one more eager member of the team to make the project sustainable. "making a release" is a hostage. I can use this card only once.

But I don't think that "card" works particularly well. It's not easy for people to find this project, and very few will be interested in seriously digging into it or co-maintaining it, if it's fairly esoteric and complicated, and works pretty damn well already. Most users aren't that interested in the detailed documentation, they just want the basics to generate their tags for a popular traditional language.

The only real chance is as follows: make a release (call it 5.9.1 alpha if that helps). This will become the default ctags everywhere in about a year (for the relatively small set of people who still use ctags). Most users will just enjoy better ctags. Some of those people will try to use universal-ctags for less-traditional languages, find some low-hanging fruit for improvement, and come here to contribute fixes. Some of that tiny fraction of people contributing fixes for the more experimental stuff, may increase their involvement and end up helping you maintain the project. But you need to cast a super wide net to have any slim chance of getting another maintainer, and you need to tag a version (any kind of version) to do that.

As you feared, you will also get lots of random people just asking questions, all kinds of questions, mostly inappropriate for this repo. They won't be reduced at all, if the release is bug-free and fully documented - there will be just as many.

-----

In the meantime, like a few others, I've been marking my own unofficial versions, just to more conveniently track which commits I've used, and what I'm currently running. But I also have the audacity to number them "5.99.x" and post the link here: https://github.com/ploxiln/ctags/releases

Fwiw, in my fake personal-convenience beta releases, I have carefully updated just the version number in the test suite expected outputs, checked that the test suite passes on both macOS and Linux, and I have generated dist tarballs including the configure script. Maybe these are useful to some people, maybe not really as they are unofficial, maybe they give a demonstration of how little effort is expected beyond just picking a number and making a tag ...

**Comment by masatake** [comment]
I added a man page that shows how the python parser of ctags works.
 docs/man/ctags-lang-python.7.rst in 
https://github.com/universal-ctags/ctags/pull/2613/files#diff-ace8fdcd38956865f8b696fd2a0efe6b

Could you review the change?

**Comment by alerque** [comment]
This issue saga might be the most blatant case study for [Perfect is the enemy of good](https://en.wikipedia.org/wiki/Perfect_is_the_enemy_of_good) I've seen in years. As the current maintainer of the VIM [tagbar](https://github.com/majutsushi/tagbar/) plugin that has gradually migrated from supporting both Exuberant and Universal variants to only supporting Universal Ctags I've many times chased my tail because while this project has already become the *de facto* best tag engine to run, without tagged releases the end user situation is just chaos.

Please stop making excuses about this not being perfect and start tagging releases. Save the big 6.0.0 for when you feel it's perfect if you want, but start using semantic release tags. I don't care if they just step though 0.0.1–0.0.999 for the foreseeable future to be upgraded to some significant version when you come up with a scheme you are happy with and all the parts, but occasionally tag this as something at intervals where you are using it yourself and know at least the basics work.

And yes If you want help I'll help automate the release process. I have a lot of experience with build systems and have build automated release pipelines where just adding the git tag is enough to cut a release, build, and post the release artifacts. It's not that hard, and I (and probably others) will pitch in to help, but you have to stop blocking the idea.

**Comment by masatake** [comment]
Did you see https://github.com/universal-ctags/ctags-win32 run by @k-takata?
I don't know the mechanism but the repository puts a tag automatically when we made commits to the main ctags repository. In ctags-win32 also makes a binary package for windows after putting a tag. However, we don't have an interest in this stage.

1. create a new repository named ctags-release in the universal-ctags organization by cloning the main ctags repository.
    You are the maintainer of the project.

2. When the main ctags repository is updated, update ctags-release by using the technique used in ctags-win32 repo.

3. Put a tag any time you want. I think the approach taken by ctags-win32 repo is the best. However, it is up to you.

How do you think about this idea? This idea doesn't take any time of mine (this is good for me), and this approach doesn't depend on my  "Perfect is the enemy of good" attitude (this is good for people other than I). You can decide how frequently tagging and the version name scheme.

If possible I would like you to work with universal-ctags/ctags-win32 and universal-ctags/homebrew-universal-ctags to use the same version name scheme. 

> And yes If you want help I'll help automate the release process. I have a lot of experience with build systems and have build automated release pipelines where just adding the git tag is enough to cut a release, build, and post the release artifacts. It's not that hard, and I (and probably others) will pitch in to help, but you have to stop blocking the idea.

Thank you very much. I would like you to consider the above idea.

**Comment by k-takata** [comment]
> I don't know the mechanism but the repository puts a tag automatically when we made commits to the main ctags repository.

ctags-win32 uses a very simple way. It checks the universal-ctags repository at 0:00 UTC every day, and if any new commits are found, creates a new tag.
The tag is created by the following commands:
```shell
cd ctags
ctagsver=$(git describe --tags --always)    # Get the tag name if it exists, otherwise get the commit ID.
cd ..
git tag $(date --rfc-3339=date)/$ctagsver   # Use YYYY-MM-DD/CommitID format.
```
(See https://github.com/universal-ctags/ctags-win32/blob/master/scripts/update-repo.sh)

To be honest, I'm also waiting for a release.
My mind is not so changed from last year: https://github.com/universal-ctags/ctags/issues/1038#issuecomment-550633120

(From your old comment)
> What I'm afraid of is my burning out. As this issue shows, so many people want a release.
> (e-)ctags is a really big name, so once we make a release, many people may use the version of u-ctags.

Burning out should be really avoided, but I doubt this. Many people are already using u-ctags without knowing its actual version (like #2641 or https://github.com/universal-ctags/ctags/issues/1038#issuecomment-653189529).
Creating a release may not increase the users so much. And there is an obvious befit for you. You will not suffer from the requests to create a release. ;-)

**Comment by masatake** [comment]
@k-takata, how do you run the script?
Is it possible to run the script here, the main ctags repository?

> To be honest, I'm also waiting for a release.

If we run the script here, and the script puts tags, can I say I provide the releases that you want?

**Comment by k-takata** [comment]
ctags-win32 uses AppVeyor's cron to run the script.
GitHub Actions also supports scheduled builds.

> If we run the script here, and the script puts tags, can I say I provide the releases that you want?

Hmm, we should carefully consider the versioning rule.
I prefer [semver](https://semver.org/). E.g. `v6.0.0-alpha-YYYYMMDD`.

However, I know a project that doesn't follow semver: https://github.com/google/re2
They use the `YYYY-MM-DD` style and create a tag once a month.

**Comment by codebrainz** [comment]
Github also has "releases" feature as @ploxiln uses in their fork, it takes like 30 seconds to make a "release". The form looks like this:

<img width="486" alt="gh_release" src="https://user-images.githubusercontent.com/181177/92824880-c9323e80-f383-11ea-93a9-1ce1f9aada1f.png">

You can use the "This is a pre-release" to signify to people that it's not a stable release. If you put the next version number in the form, it will make the Git tag automatically. You also don't have to add build artifacts since it will automatically create a zip/tarball from the repo, though it might be useful to provide a `make dist` tarball so people don't need Autotools and stuff to build with it.

**Comment by masatake** [comment]
@k-takata, thank you for responding.

Supporting semver in this stage is hard. 

In the summary of semver:
> MAJOR version when you make incompatible API changes,

This means I have to manage MAJOR version carefully.
People say just tagging is enough. semver is more than just tagging.

My idea is v5.9.YYYYMMDD. I don't want to take care of API changes when tagging.

@codebrainz, thank you for your input. However, I don't want to make an archive.
If the Geany development process needs it, I will make it. What I should do is merge the parsers in the pull request from Geany project.

**Comment by ploxiln** [comment]
> My idea is v5.9.YYYYMMDD. I don't want to take care of API changes when tagging.

> @codebrainz, thank you for your input. However, I don't want to make an archive.

Fair, and appreciated, thanks.

**Comment by codebrainz** [comment]
> @codebrainz, thank you for your input. However, I don't want to make an archive.
> If the Geany development process needs it, I will make it. What I should do is merge the parsers in the pull request from Geany project.

I wasn't really speaking as a Geany developer, I was just giving a suggestion/info on a simple way to make people who want a "release" happy now without burdening you too much, by just using the built-in Github "release" mechanism to cut a release.

I'm not versioning expert or package maintainer, but it seems kind of pointless to auto-generate tags/releases for the purpose of this Issue since the tags/releases are as arbitrary as just using Git SHAs of a commit and everyone will be using different auto-tagged versions.

**Comment by masatake** [comment]
> I wasn't really speaking as a Geany developer, I was just giving a suggestion/info on a simple way to make people who want a "release" happy now without burdening you too much, by just using the built-in Github "release" mechanism to cut a release.
 
I see. Thank you.

**Comment by ploxiln** [comment]
As a small complication, I _would_ like for `ctags --version` output to correspond to the git version tag ... and this version ends up in lots of test output unfortunately ... so ideally there is a little bit more to it than just an automated git tag. But anyway, I'll take whatever I can get, even just a git tag is better than nothing :)

(for a quick overview of _everywhere_ the version appears, you can take a look at https://github.com/universal-ctags/ctags/compare/b599093...ploxiln:9363e33)

**Comment by masatake** [comment]
> As a small complication,

Understandable. But I focus only on tagging at this time.

@k-takata, is the naming scheme for tagging "v5.9.YYYYMMDD" acceptable?

**Comment by docwhat** [comment]
Semver may be better thought of as BREAKING, FEATURE, FIX instead of MAJOR, MINOR, PATCH.

I don't have much experience, but is detecting BREAKING changes hard when writing a commit?

That being said, I'm a beggar here and not a chooser.

Even a regular date-stamp style tag is better than nothing.  As long as a human or computer can tell which is newer.

**Comment by k-takata** [comment]
That might be acceptable even it doesn't follow semver strictly.
Are you thinking of tagging automatically or manually?

**Comment by masatake** [comment]
@docwhat, I know semver is much better.  However, in 5.9.x series, I will not choose the semver scheme.
From 6.x, I will choose the scheme, however, I cannot imagine when 6.x is available.
I wrote various reasons I don't make a release. I will add one here.

I found there are two levels in ctags about semver.

* the main part level

   the command line interface like --regex-<LANG>, --list-kinds, ... are included here.
   Most of my work in this project is about the main part.
   As @docwhat wrote, I recognize I'm dong "BREAKING" or not.
   Therefore I use "[SELF-INCOMPATIBLE]" in the commit log. It implies the need of a major version 
   up in the semver meaning. About the main part, it is possible to take the semver scheme.

* the parsers' part level

   A parser defines kinds, fields, and extras. I would like to avoid BREAKING however, I made and I accepted
   BREAKING. I think I have to introduce more breaking especially about  #2428.
   I wonder such breaking in a parser level needs incrementing MAJOR or not.
   Consider you are a C programmer. If we introduce a "BREAKING" change in our COBOL parser, 
   do you think u-ctags introduces a "BREAKING" change? I guess you don't care.

I think the semver should be applied to u-ctags in the above two levels separately.
So when referring a version of u-ctags, you (a C progammer) may say:

 u-ctags-6.2.1 + c-parser-1.2.0

A coble programmer may say
 
 u-ctags-6.2.1 + cobol-parser-1.0.3

I wonder how to represent this situation in the semver naming scheme.

**Comment by masatake** [comment]
@k-takata, I'm thinking about automatically.
And as usual, I need your help.
Can I ask you to do the similar as you did in ctags-win32?

**Comment by polyzen** [comment]
For handling of the date part of the version, this could provide insight:
https://calver.org/

**Comment by docwhat** [comment]
If you are building a package made of parts with language parsers and a core engine then you would have a version number for each parser (e.g. cobol-parser 1.0.3), a version for the core engine (core-engine 6.2.1) and then a "combined" package number: u-ctags 6.3.0

If any parser or the engine bump a major number, then the combined package would bump its major number too.

Yes, that means that the combined package may reach triple digits quickly, but that's okay.  The version numbers are just numbers.

Alternatively, you can manage the combined package version entirely as "marketing"  ... e.g. with a date only (via calver.org schemes)

Both are legit.

**Comment by masatake** [comment]
I will implement weekly-automatic-tagging v5.9.YYYYMMDD.0 as trial.
I appended ".0" at the end for allowing us to release an extra version manually.

**Comment by masatake** [comment]
See #2655.

**Comment by masatake** [comment]
The initial tag,  p5.9.20201011.0, was created.

**Comment by polyzen** [comment]
Does the `p` at the beginning of the version have any meaning?

**Comment by masatake** [comment]
> Edit: Seems it denotes "periodical"?

Maybe yes. I cannot remember the meaning. I was looking for a character other than 'v'.

**Comment by masatake** [comment]
I'm thinking about releasing 6.0. I put "ctags6" label in our issues and pull requests.
